lib/utils.js

```diff
@@ -229,6 +229,144 @@ const format = body => {
   return data;
 };
 
+/*
+ * @typedef Reference
+ * @type object
+ *
+ * @property {type} Reference.id - Id of the content type
+ * @property {attr} Reference.attr - Attribute Index
+ * @property {input} Reference.input - Input of the reference
+ * @property {length} Reference.length - Number of instances
+ * @property {ct.id} Reference.ct.id - Id of the referenced content type
+ * @property {ct.index} Reference.ct.index - Position of the reference content type
+ */
+
+/*
+ * Inititalizes reference
+ *
+ * @param {object} types - content types object
+ *
+ *
+ * @returns {array[Reference]} - content type with reference data
+ */
+const references = (types) => {
+  const cts = _.cloneDeep(types);
+  const refs = [];
+  const position = {};
+
+  // Saves position of each content type
+  cts.forEach((type, index) => {
+    if (!position.hasOwnProperty(type.id)) {
+      position[type.id] = index;
+    }
+  });
+
+  // Iterates through each content type and attributes
+  cts.map(ct => {
+    ct.attributes.map((attr, index) => {
+      let inputs;
+      let length;
+
+      // Get the number of instances if repeatable
+      if (Array.isArray(attr.inputs)) {
+        inputs = attr.inputs[0];
+        length = attr.inputs.length;
+      }
+      else {
+        inputs = attr.inputs;
+      }
+
+      // Iterate through each inputs
+      Object.keys(inputs).forEach(input => {
+        // Checks if input is a reference
+        if (inputs[input].hasOwnProperty('reference') && inputs[input].reference === true) {
+          // Throw error if contentType is not defined
+          if (!inputs[input].settings.hasOwnProperty('contentType')) {
+            throw new Error('Reference must have a content type');
+          }
+
+          // Throw error if contentType is not valid
+          if (!position.hasOwnProperty(inputs[input].settings.contentType)) {
+            throw new Error(`Content Type ${inputs[input].settings.contentType} is not valid`);
+          }
+
+          // Defines and push reference object
+          refs.push({
+            type: ct.id,
+            attr: index,
+            input,
+            length,
+            ct: {
+              index: position[inputs[input].settings.contentType],
+              id: inputs[input].settings.contentType,
+            },
+          });
+        }
+      });
+    });
+  });
+
+  return refs;
+};
+
+
+/*
+ * Fills in content type data from database
+ *
+ * @param {object} types - content types object
+ * @param {string} id - id of the requested content type
+ * @param {array} refs - reference data
+ * @param {object} database - database object
+ *
+ *
+ * @returns {object} - content type object with reference data
+ */
+const fill = (types, id, refs, database) => {
+  const ct = singleItem('id', id, types);
```

Comment:
we were calling `singleItem` to get the type.
So, we can either call this function before `fill` or here.
Will not make any difference